Lib/test/test_type_params.py

```diff
@@ -0,0 +1,563 @@
+        run_code("""def func[A](A): ...""")
+
+    def test_name_non_collision_03(self):
+        run_code("""def func[A](*A): ...""")
+
+    def test_name_non_collision_04(self):
+        # Mangled names should not cause a conflict.
+        run_code("""\
+            class ClassA:
+                def func[__A](self, __A): ...
+            """
+        )
+
+    def test_name_non_collision_05(self):
+        run_code("""\
+            class ClassA:
+                def func[_ClassA__A](self, __A): ...
+            """
+        )
+
+    def test_name_non_collision_06(self):
+        run_code("""\
+            class ClassA[X]:
+                def func(self, X): ...
+            """
+        )
+
+    def test_name_non_collision_07(self):
+        run_code("""\
+            class ClassA[X]:
+                def func(self):
+                    X = 1
+            """
+        )
+
+    def test_name_non_collision_08(self):
+        run_code("""\
+            class ClassA[X]:
+                def func(self):
+                    a = [X for X in []]
+            """
+        )
+
+    def test_name_non_collision_9(self):
+        run_code("""\
+            class ClassA[X]:
+                def func[X](self):
+                    ...
+            """
+        )
+
+    def test_name_non_collision_10(self):
+        run_code("""\
+            class ClassA[X]:
+                X: int
+            """
+        )
+
+    def test_name_non_collision_11(self):
+        run_code("""\
+            def outer():
+                X = 1
+                def inner[X]():
+                    nonlocal X
+            """
+        )
+
+    def test_name_non_collision_13(self):
+        run_code("""\
+            X = 1
+            def outer():
+                def inner[X]():
+                    global X
+            """
+        )
+
+    def test_disallowed_expressions(self):
+        with self.assertRaises(SyntaxError):
+            run_code("type X = (yield)")
+        with self.assertRaises(SyntaxError):
+            run_code("type X = (yield from x)")
+        with self.assertRaises(SyntaxError):
+            run_code("type X = (await 42)")
+        with self.assertRaises(SyntaxError):
+            run_code("async def f(): type X = (yield)")
+        with self.assertRaises(SyntaxError):
+            run_code("type X = (y := 3)")
+        with self.assertRaises(SyntaxError):
+            run_code("class X[T: (yield)]: pass")
+        with self.assertRaises(SyntaxError):
+            run_code("class X[T: (yield from x)]: pass")
+        with self.assertRaises(SyntaxError):
+            run_code("class X[T: (await 42)]: pass")
+        with self.assertRaises(SyntaxError):
+            run_code("class X[T: (y := 3)]: pass")
+        with self.assertRaises(SyntaxError):
+            run_code("class X[T](y := Sequence[T]): pass")
+        with self.assertRaises(SyntaxError):
+            run_code("def f[T](y: (x := Sequence[T])): pass")
```

Comment:
Could also consider using `test.support.check_syntax_error`:

https://github.com/python/cpython/blob/85c7bf5bcec07beea6064976e6199195cd34329d/Lib/test/support/__init__.py#L622-L631

Example usage:

https://github.com/python/cpython/blob/85c7bf5bcec07beea6064976e6199195cd34329d/Lib/test/test_positional_only_arg.py#L25-L45

Reply:
Good idea, just pushed that.